tooling/Microsoft.VisualStudio.BlazorExtension/AutoRebuild/BuildEventsWatcher.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.BlazorExtension
+{
+    /// <summary>
+    /// Watches for Blazor project build events, starts new builds, and tracks builds in progress.
+    /// </summary>
+    internal class BuildEventsWatcher : IVsUpdateSolutionEvents2
+    {
+        private const string BlazorProjectCapability = "Blazor";
+        private readonly IVsSolution _vsSolution;
+        private readonly IVsSolutionBuildManager _vsBuildManager;
+        private readonly object mostRecentBuildInfosLock = new object();
+        private readonly Dictionary<string, BuildInfo> mostRecentBuildInfos = new Dictionary<string, BuildInfo>();
+
+        public BuildEventsWatcher(IVsSolution vsSolution, IVsSolutionBuildManager vsBuildManager)
+        {
+            _vsSolution = vsSolution ?? throw new ArgumentNullException(nameof(vsSolution));
+            _vsBuildManager = vsBuildManager ?? throw new ArgumentNullException(nameof(vsBuildManager));
+        }
+
+        public Task<bool> PerformBuildAsync(string projectPath, DateTime allowExistingBuildsSince)
+        {
+            BuildInfo newBuildInfo;
+
+            lock (mostRecentBuildInfosLock)
+            {
+                if (mostRecentBuildInfos.TryGetValue(projectPath, out var existingInfo))
+                {
+                    // If there's a build in progress, we'll join that even if it was started
+                    // before allowExistingBuildsSince, because it's too messy to cancel
+                    // in-progress builds. On rare occasions if the user is editing files while
+                    // a build is in progress they *might* see a not-latest build when they
+                    // reload, but then they just have to reload again.
+                    var acceptBuild = !existingInfo.TaskCompletionSource.Task.IsCompleted
```

Comment:
I was thinking that, in the future, for failed builds we could capture the build error message and that would be stored on the task result. Then we'd use that task and be able to show the build error in the browser. The task would get replaced when the developer edits a file again and reloads, or manually triggers a build. So it was intentional to accept failed builds here if they are the most "up to date" reflection of the state of your app.

Does that seem reasonable, or do you think the dev experience would be better if we changed this? I thought it was pointless to request another build until the source code has been edited again.

Reply:
So for now, the experience when you have an error will be that we just don't get new content? I guess some things might be updated and some not.

I think this is good enough for preview2.

Reply:
Yes. Typically things won't have updated after a build failure because the compiled assembly represents the last successful build.

Later on I do think we should try to get the build errors to show up in the browser, otherwise this might be confusing.